File: ios/CustomUIViewRegistry.swift
import UIKit

struct CustomUIStyle {
  let visibility: String?
  let backgroundColor: String?
}

final class CustomUIViewRegistry {
  static let shared = CustomUIViewRegistry()
  private var stylePolicies: [String: CustomUIStyle] = [:]

  private init() {}

  private func policyKey(viewType: String, theme: String?) -> String {
    guard let theme, !theme.isEmpty else { return viewType }
    return "\(viewType)#\(theme.lowercased())"
  }

  func setStyle(viewType: String, visibility: String?, backgroundColor: String?, theme: String?) {
    let key = policyKey(viewType: viewType, theme: theme)
    let existing = stylePolicies[key]
    let merged = CustomUIStyle(
      visibility: visibility ?? existing?.visibility,
      backgroundColor: backgroundColor ?? existing?.backgroundColor
    )
    stylePolicies[key] = merged
  }

  func clearStyle(viewType: String, theme: String?) {
    stylePolicies.removeValue(forKey: policyKey(viewType: viewType, theme: theme))
  }

  func style(viewType: String, theme: String?) -> CustomUIStyle? {
    if let theme {
      if let themed = stylePolicies[policyKey(viewType: viewType, theme: theme)] {
        return themed
      }
    }
    return stylePolicies[policyKey(viewType: viewType, theme: nil)]
  }

  func applyStyle(view: UIView, style: CustomUIStyle) {
    DispatchQueue.main.async {
      if let visibility = style.visibility {
        view.isHidden = (visibility == "hidden")
      }
      if let backgroundColor = style.backgroundColor, let color = self.colorFromHex(backgroundColor) {
        view.backgroundColor = color
      }
    }
  }

  private func colorFromHex(_ value: String) -> UIColor? {
    var hex = value.trimmingCharacters(in: .whitespacesAndNewlines)
    if hex.hasPrefix("#") {
      hex.removeFirst()
    }
    if hex.count == 6 {
      hex = "FF" + hex
    }
    guard hex.count == 8, let hexNumber = UInt64(hex, radix: 16) else {
      return nil
    }
    let alpha = CGFloat((hexNumber & 0xFF000000) >> 24) / 255.0
    let red = CGFloat((hexNumber & 0x00FF0000) >> 16) / 255.0
    let green = CGFloat((hexNumber & 0x0000FF00) >> 8) / 255.0
    let blue = CGFloat(hexNumber & 0x000000FF) / 255.0
    return UIColor(red: red, green: green, blue: blue, alpha: alpha)
  }
}

File: android/src/main/java/com/viafoura/reactnative/CustomUIViewRegistry.kt
package com.viafoura.reactnative

import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.view.View

data class CustomUIStyle(
  val visibility: String? = null,
  val backgroundColor: String? = null
)

object CustomUIViewRegistry {
  private val stylePolicies = mutableMapOf<String, CustomUIStyle>()

  private fun policyKey(viewType: String, theme: String?): String {
    val normalizedTheme = theme?.lowercase()
    return if (normalizedTheme.isNullOrBlank()) {
      viewType
    } else {
      "$viewType#$normalizedTheme"
    }
  }

  @Synchronized
  fun setStyle(viewType: String, style: CustomUIStyle, theme: String?) {
    val key = policyKey(viewType, theme)
    val existing = stylePolicies[key]
    val merged = CustomUIStyle(
      visibility = style.visibility ?: existing?.visibility,
      backgroundColor = style.backgroundColor ?: existing?.backgroundColor
    )
    stylePolicies[key] = merged
  }

  @Synchronized
  fun clearStyle(viewType: String, theme: String?) {
    stylePolicies.remove(policyKey(viewType, theme))
  }

  @Synchronized
  fun getStyle(viewType: String, theme: String?): CustomUIStyle? {
    val themed = stylePolicies[policyKey(viewType, theme)]
    return themed ?: stylePolicies[policyKey(viewType, null)]
  }

  fun applyStyle(view: View, style: CustomUIStyle) {
    val runnable = Runnable {
      style.visibility?.let { value ->
        view.visibility = if (value == "hidden") View.GONE else View.VISIBLE
      }
      style.backgroundColor?.let { colorValue ->
        parseColor(colorValue)?.let { parsedColor ->
          view.setBackgroundColor(parsedColor)
        }
      }
    }
    if (Looper.myLooper() == Looper.getMainLooper()) {
      runnable.run()
    } else {
      Handler(Looper.getMainLooper()).post(runnable)
    }
  }

  private fun parseColor(value: String): Int? {
    return try {
      Color.parseColor(value)
    } catch (error: IllegalArgumentException) {
      null
    }
  }
}
